pkg/trait/builder.go

```diff
@@ -56,9 +57,18 @@ func (t *builderTrait) InfluencesKit() bool {
 	return true
 }
 
-// InfluencesBuild overrides base class method.
-func (t *builderTrait) InfluencesBuild(this, prev map[string]interface{}) bool {
-	return true
+func (t *builderTrait) Matches(trait Trait) bool {
+	otherTrait, ok := trait.(*builderTrait)
+	if !ok {
+		return false
+	}
+	if t.BaseImage != otherTrait.BaseImage || len(t.Properties) != len(otherTrait.Properties) {
+		return false
+	}
+	// More sofisticated check if len is the same. Sort and compare via slices equal func.
+	slices.Sort(t.Properties)
+	slices.Sort(otherTrait.Properties)
```

Comment:
nitpick: this method has a side effect as it alters the order of both the current and the other trait properties.

Reply:
Yes, but in this case I don't think it's a problem. This func is not directly used by the trait per se but is a supporting function used in the `matchesComparableTrait` func. Here what we do is to create 2 brand new traits which are only used for comparison, so, I tend to think they are not really affecting the rest of the execution as they are 2 new objects:

https://github.com/apache/camel-k/pull/5230/files#diff-80f4821d91dcccf8c11c32962f56b41b7dcfccadeb6366df3cf49baa55188ba1R637-R658

I think the structure backing the two object is different. But I may be wrong. If you feel we need to prove it I can try to develop a unit test to make sure this is not happening. Thanks.

Reply:
The structure backing the two object is definitively different, my point was more on the fact if `Matches` is or will be used concurrently, then it may result in an undefined behavior.

Given the current PR I don't think it can happens but since the method is a public method part of some of the traits, then it may be used also outside the current scope and I did want to raise it to avoid falling in the same trap as  https://github.com/apache/camel-k/pull/5188#discussion_r1512280139

Reply:
Yes, I see. For the sake of making the thing in a clean way, I can copy the array and sorting/comparing them. It should not be too heavy from a computation pov and it will save us against remote (but still possible) inconsistencies.

Reply:
That would be a sensible option.

Another one, but would require some more work so I don't think it is worth at this stage, could be to do something similar what it is done in the golang standard library where there is often the possibility to pass a function for equality, sorting, etc. like https://pkg.go.dev/slices#EqualFunc